core/src/main/scala/org/apache/spark/SparkContext.scala

```diff
@@ -1489,7 +1489,7 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
   private[spark] def killAndReplaceExecutor(executorId: String): Boolean = {
     schedulerBackend match {
       case b: CoarseGrainedSchedulerBackend =>
-        b.killExecutors(Seq(executorId), replace = true)
+        b.killExecutors(Seq(executorId), replace = true, force = false)
```

Comment:
Why is this one `force = false`? This is most probably the cause of the failure in the heartbeat test; if the executor is not heartbeating, the driver should forcefully kill it, regardless of whether tasks are currently assigned to it.